### PR TITLE
Added Dockerfile to individual components

### PR DESCRIPTION
**Pull request by jpkrohling**

### CODE REVIEW

**Comment by jpkrohling** [comment]
This PR adds the following `make` targets:

* `docker` -- to build the Docker images locally, based on the newly created Dockerfiles. It does *not* build docker images for the existing images, as those are handled by other tasks (all-in-one and crossdock)
* `docker-push` -- to push the images built by the `docker` target to a repository. This requires a previous `docker login`.

The Travis changes were made kind of in a blind mode, as I'm not sure how I can test it. The bash parts were tested as much as I could.

I'm also not quite sure how I can remove the submodule updates to this commit, as I never really dealt with submodules. Any hints there is highly appreciated :)

File: plugin/storage/cassandra/schema.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
```
Comment (jpkrohling) [inline plugin/storage/cassandra/schema.sh]: Not quite sure if this is the most appropriate for this script. It has to be either on the same directory as the Dockerfile or in a sub directory, but not on a parent directory.

File: cmd/agent/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM centos:7
```
Comment (jpkrohling) [inline cmd/agent/Dockerfile]: The images are inheriting the CentOS 7 images, as those provide some tools which are useful for debugging. Once the images mature a bit on the community, we can change the base image to Alpine or similar.

File: .travis.yml [hunk 3]
```diff
@@ -16,6 +16,9 @@ matrix:
   - go: 1.7
     env:
     - CROSSDOCK=true
+  - go: 1.7
+    env:
+    - DOCKER=true
```
Comment (jpkrohling) [inline .travis.yml]: I have to admit that I don't know for sure what this does. I *guess* this will trigger different executors on Travis, but I just did a copy/paste of the sections above.

File: Makefile [hunk 4]
```diff
@@ -46,7 +47,7 @@ md-to-godoc-gen:
 
 .PHONY: clean
 clean:
-	rm -rf cover.out cover.html lint.log fmt.log
+	rm -rf cover.out cover.html lint.log fmt.log jaeger-ui-build
```
Comment (jpkrohling) [inline Makefile]: This doesn't really belong to this commit. I hope it's OK to leave this change here.

File: Makefile [hunk 5]
```diff
@@ -111,6 +112,21 @@ build-query-linux:
...
+docker: build_ui build-agent-linux build-collector-linux build-query-linux
+	cp -r jaeger-ui-build/build/ cmd/query/jaeger-ui-build
+	docker build -t $(DOCKER_NAMESPACE)/jaeger-cassandra-schema plugin/storage/cassandra/ ; \
+	for component in agent collector query ; do \
+		docker build -t $(DOCKER_NAMESPACE)/jaeger-$$component cmd/$$component ; \
+	done
+	rm -rf cmd/query/jaeger-ui-build
+
+.PHONY: docker-push
+docker-push:
```
Comment (jpkrohling) [inline Makefile]: This target is only useful for development purposes and it's not being used for the actual builds. Should this be removed?

File: plugin/storage/cassandra/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,7 @@
+FROM jpkroehling/cassandra
```
Comment (jpkrohling) [inline plugin/storage/cassandra/Dockerfile]: This was discussed on the mailing list. I'll either change this to use the "official" Dockerhub image for Cassandra or change to use an image we are building for OpenShift. I don't expect to change this before this PR is merged, though.

File: travis/build-docker-images.sh [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+set -e
+
+export BRANCH=$(if [ "$TRAVIS_PULL_REQUEST" == "false" ]; then echo $TRAVIS_BRANCH; else echo $TRAVIS_PULL_REQUEST_BRANCH; fi)
+# Only push the docker container to Docker Hub for master branch
+if [[ "$BRANCH" == "master" && "$TRAVIS_SECURE_ENV_VARS" == "true" ]]; then echo 'upload to Docker Hub'; else echo 'skip docker upload for PR'; exit 0; fi
+
+source ~/.nvm/nvm.sh
+nvm use 6
+DOCKER_NAMESPACE=jaegertracing make docker
```
Comment (black-adder) [inline travis/build-docker-images.sh]: stupid question incoming: shouldn't there be a newline before `make docker`?

File: Makefile [hunk 8]
```diff
@@ -111,6 +112,21 @@ build-query-linux:
 build-collector-linux:
 	CGO_ENABLED=0 GOOS=linux installsuffix=cgo go build -o ./cmd/collector/collector-linux ./cmd/collector/main.go
 
+.PHONY: docker
+docker: build_ui build-agent-linux build-collector-linux build-query-linux
+	cp -r jaeger-ui-build/build/ cmd/query/jaeger-ui-build
+	docker build -t $(DOCKER_NAMESPACE)/jaeger-cassandra-schema plugin/storage/cassandra/ ; \
```
Comment (black-adder) [inline Makefile]: Can we call this jaeger-cassandra instead? The schema gets applied on startup and after that it's essentially just a cassandra box. The name might throw off some people.

File: cmd/agent/Dockerfile (hunk 2 above)
Comment (jpkrohling) [inline cmd/agent/Dockerfile]: The CentOS image isn't that big (192.5 MB), which is shared by all images. But if you have a specific image to suggest, I could try that out and see if it works.

File: Makefile (hunk 5 above)
Comment (jpkrohling) [inline Makefile]: I did quite a few times under my own namespace, to test the images on OpenShift/Kubernetes. If I add a comment to this target, saying that this is intended only for local development purposes, would that be OK to leave it there?

File: Makefile (hunk 8 above)
Comment (jpkrohling) [inline Makefile]: There's no "after" for this image :) It's used as a Kubernetes "Job": it will just run the shell script that generates the keyspace and will gracefully die afterwards. It's derived from the Cassandra image because it needs the `cqlsh` tools, so, I thought it would be easier (and save space) to just consume a image that I know existed there, but Cassandra itself will never start on this image.

File: travis/build-docker-images.sh (hunk 7 above)
Comment (jpkrohling) [inline travis/build-docker-images.sh]: Not really. In bash, this has the effect of setting this var for this command, so, `make` will see DOCKER_NAMESPACE as `jaegertracing` without the parent script needing to `export` it first.

File: Makefile (hunk 5 above)
Comment (black-adder) [inline Makefile]: sgtm

File: Makefile (hunk 8 above)
Comment (black-adder) [inline Makefile]: ahhh gotcha. Maybe jaeger-cassandra-bootstrap? fine with jaeger-cassandra-schema too.

File: travis/build-docker-images.sh (hunk 7 above)
Comment (black-adder) [inline travis/build-docker-images.sh]: TIL

File: Makefile [hunk 9]
```diff
@@ -111,6 +112,21 @@ build-query-linux:
 build-collector-linux:
 	CGO_ENABLED=0 GOOS=linux installsuffix=cgo go build -o ./cmd/collector/collector-linux ./cmd/collector/main.go
 
+.PHONY: docker
+docker: build_ui build-agent-linux build-collector-linux build-query-linux
+	cp -r jaeger-ui-build/build/ cmd/query/jaeger-ui-build
```
Comment (yurishkuro) [inline Makefile]: I am not clear why this is needed. We already build all-in-one without copying this dir into query source, instead we just mount it to the Docker image from `./jaeger-ui-build`

File: Makefile (hunk 8 above)
Comment (yurishkuro) [inline Makefile]: I would prefer to use the same mechanism across all artifacts. 

@black-adder  It sounds like this is doing something different from what e2e test is doing to init Cassandra.

File: cmd/collector/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,6 @@
+FROM centos:7
```
Comment (yurishkuro) [inline cmd/collector/Dockerfile]: I thought we'd put all Docker files in one dir. But no strong feeling either way.

File: cmd/collector/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+EXPOSE 14267
+
+COPY collector-linux /go/bin/
+
+CMD ["/go/bin/collector-linux"]
```
Comment (yurishkuro) [inline cmd/collector/Dockerfile]: This comment applies to all images - don't we need to be translating env vars into command line switches, to make the images follow https://12factor.net/ ?

Ideally our binaries could be reading env vars directly, there are frameworks like spf13/Cobra that support that, but we haven't implemented it yet. So the only way currently to configure each binary is to translate env vars into command line switches.

File: plugin/storage/cassandra/schema.sh [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    echo "Cassandra is still not up at ${CQLSH_HOST}. Waiting 1 second."
+    sleep 1s
+    ((total_wait++))
+  fi
+done
+
+echo "Generating the schema for the keyspace ${KEYSPACE} and datacenter ${DATACENTER}"
+export KEYSPACE
+
+# the `test` parameter is to force the script to use a SimpleStrategy instead of
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: why? simple strategy only makes sense in test env with a single Cassandra host.

File: plugin/storage/cassandra/schema.sh (hunk 1 above)
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: looks ok here, given the other scripts. I would just rename it to `create-schema.sh`

File: plugin/storage/cassandra/schema.sh [hunk 13]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+
+CQLSH_HOST=${CQLSH_HOST:-"cassandra"}
+CASSANDRA_WAIT_TIMEOUT=${CASSANDRA_WAIT_TIMEOUT:-"60"}
+DATACENTER=${DATACENTER:-"openshift"}
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: this doesn't look like a good default. I am actually not sure what datacenter the official Cassandra image mentions.

File: plugin/storage/cassandra/schema.sh [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+KEYSPACE=${KEYSPACE:-"jaeger_v1_${DATACENTER}"}
+
+total_wait=0
+while true
+do
+  ping -c 1 ${CQLSH_HOST} > /dev/null 2>&1
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: is this the typical way of checking that the container app is running? Seems line it might succeed as soon as the container / IP is available, but the Cassandra server may still be starting. 

@black-adder what did we use for e2e test? Didn't we wait for the actual response from Cassandra?

File: Makefile (hunk 8 above)
Comment (black-adder) [inline Makefile]: What we do for e2e is pretty ridiculous though, we keep bringing up collector and wait until it doesn't fatal out due to cassandra not being ready. I prefer this new approach.

File: Makefile (hunk 9 above)
Comment (jpkrohling) [inline Makefile]: That's because the Dockerfile is within the cmd/query , and Docker can only build stuff that is on its own directory or any sub directory, but not in any parent directory (for security reasons I suppose). So, we need to have a copy there.

File: cmd/collector/Dockerfile (hunk 10 above)
Comment (jpkrohling) [inline cmd/collector/Dockerfile]: We could, as long as the contents for the Docker images are in sub directories from where the Dockerfiles are located. I'd rather keep it alongside with the image's contents.

File: cmd/collector/Dockerfile (hunk 11 above)
Comment (jpkrohling) [inline cmd/collector/Dockerfile]: Not necessarily: I'm not quite sure how it would work for Docker Compose, but for Docker at the command line, Kubernetes and OpenShift, you can override the command line and set extra env vars.

So, the best approach is to specify only what's really required to get the binary running, so that you don't need to maintain two places with default values.

```bash
$ docker run -e MY_ENV_VAR=test -it jpkroehling/jaeger-agent bash
[root@86fc09d20bfc /]# echo $MY_ENV_VAR
test
```

File: plugin/storage/cassandra/schema.sh (hunk 13 above)
Comment (jpkrohling) [inline plugin/storage/cassandra/schema.sh]: The datacenter is actually required by the existing schema generator: https://github.com/uber/jaeger/blob/master/plugin/storage/cassandra/cassandra3v001-schema.sh#L23

So, it's either `prod` or `test`, and when passing `prod`, a datacenter has to be provided.

File: plugin/storage/cassandra/schema.sh (hunk 12 above)
Comment (jpkrohling) [inline plugin/storage/cassandra/schema.sh]: I was actually using the `prod`, but got into a Cassandra issue. Something like `Error reading service_names from storage: Cannot achieve consistency level LOCAL_ONE`.

File: plugin/storage/cassandra/schema.sh (hunk 13 above)
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: Yes, it's required for `prod` because it uses the name when defining replication strategy. That datacenter must match what Cassandra uses in its `cassandra-rackdc.properties`.

File: plugin/storage/cassandra/schema.sh (hunk 12 above)
Comment (yurishkuro) [inline plugin/storage/cassandra/schema.sh]: `prod` configuration defines replication factor 2, so you need at least 2 Cassandra nodes.

File: Makefile (hunk 8 above)
Comment (jpkrohling) [inline Makefile]: I talked to @jsanda last week about something related to this and he made me realize that this might also not be appropriate.

In any case: I think we could add a separate task for having a more reliable bootstrap method, perhaps consistent among all deployment architectures. In the meantime, I'd keep it like this, so that we can move forward.

File: plugin/storage/cassandra/schema.sh (hunk 14 above)
Comment (jpkrohling) [inline plugin/storage/cassandra/schema.sh]: Kubernetes/OpenShift will establish the connection only once the target service has a container passing the "readiness" check. We don't have such a check right now, but this is more like a hack and might eventually fail on the situation you just mentioned.

I think @jsanda can help us here in coming up with a reliable solution, but I'd prefer to have it as part of another task.

File: plugin/storage/cassandra/schema.sh (hunk 12 above)
Comment (jpkrohling) [inline plugin/storage/cassandra/schema.sh]: Ok, that's it then. I'm not that familiar with Cassandra and noticed that I could change the script to `test` and it would work :-) My deployment script creates 2 Cassandra nodes, but it seems it was still not enough to avoid this error.

Should we move all these Cassandra-related tasks to a new task?

**Comment by jpkrohling** [comment]
@yurishkuro , @black-adder , I just updated this PR, rebasing it on the latest master and adding a couple of fixes based on the comment reviews. If this looks good, I'll squash it and it would be ready for merging.

There are still a couple of items open for Cassandra, mostly related to tuning it to be more resilient and reliable (replication factor, schema creation, ...) , but I'd like to have that as another follow up task, instead of blocking this PR. I'll then ask @jsanda for help on that task, as he's far more experienced with Cassandra than I am :-)

**Comment by yurishkuro** [comment]
* I am not clear why the build failed, no output in the Travis logs.
 * By default the new images are published to `jaegertracing/jaeger-{component}`, I assume we need to create repos on Docker hub?
 * The images are missing any sort of tags, usually we at least add `-t $REPO:$COMMIT`
 * Can we exclude changes to jaeger-ui from this PR?

**Comment by jsanda** [comment]
You will certainly run into problems by just pinging the the pod. Cassandra start up times will vary, particularly when you take commit log replay into consideration. That can slow down start up significantly. I suggest using ```nodetool statusbinary```.  You can grep the output for ```running```.

To avoid overly long start up times and to avoid commit log corruption, I strongly recommend running ```nodetool drain``` on shutdown.

**Comment by jpkrohling** [comment]
I updated the PR to use another command for checking Cassandra's state. I decided not to use `nodetool`, as it would require exposing the JMX port just for this purpose. If we do need JMX in the future, that's fine, but it sounded too much for this specific purpose.

> I am not clear why the build failed, no output in the Travis logs.

It works now. I had to add a dedicated conditional on `.travis.yml/install` for `$DOCKER`, instead of using a `||` on the `$ALL_IN_ONE` part.

> By default the new images are published to jaegertracing/jaeger-{component}, I assume we need to create repos on Docker hub?

They are created on demand on the first build, provided the account has push permissions on the organization.

> The images are missing any sort of tags, usually we at least add -t $REPO:$COMMIT

It's actually done on the `upload-to-docker.sh` script, called at the end of the script. It's the same script the other ones execute.

> Can we exclude changes to jaeger-ui from this PR?

I tried every git trick I know, but the only thing that helped was to do a diff against master, create a new branch off of master and apply the new diff onto this new branch :/ 

I also opened #175 to track the points to improve in our usage of Cassandra.

**Comment by jpkrohling** [comment]
Rebased with the latest master. This is ready for a re-review or eventual merge.
